File: 프로그래머스/unrated/258707. n ＋ 1 카드게임/n ＋ 1 카드게임.java
import java.util.*;


class Solution {
    static int N, coinLeft, answer = 1, idx, ans = 0;
    static Set<Integer> hand = new HashSet<>(), draw = new HashSet<>();
    public int solution(int coin, int[] cards) {
        init(cards, coin);
        return proceedGame(cards);
    }
    
   static void init(int[] cards, int coin) {
        N = cards.length + 1;
        idx = cards.length / 3;
        coinLeft = coin;
        for (int i = 0; i < idx; i++) {
            hand.add(cards[i]);
        }
    }
    static int proceedGame(int[] cards){
        int round = 1;
        while(idx + 2 <= cards.length){
            draw.add(cards[idx++]);
            draw.add(cards[idx++]);
            if(search(hand, hand)){
                round++;
                continue;
            }
            if(coinLeft > 0 && search(hand, draw)){
                coinLeft--;
                round++;
                continue;
            }
            if(coinLeft > 1 && search(draw, draw)){
                coinLeft-=2;
                round++;
                continue;
            }
            break;
        }
        return round;
    }
    
    static boolean search(Set<Integer> a, Set<Integer> b){
        for(int num : a){
            if(b.contains(N - num)){
                b.remove(N - num);
                a.remove(num);
                return true;
            }
        }
        return false;
    }
}

File: 프로그래머스/3/258707. n ＋ 1 카드게임/n ＋ 1 카드게임.py
n = 0
def solution(coin, cards):
    global n
    possible = True
    n = len(cards)
    rounds = 1
    
    hand = cards[:n // 3]
    cards = cards[n//3:]
    draw = []
    
    while cards:
        draw += cards[:2]
        cards[:] = cards[2:]
        #print(f"draw = {draw}, cards = {cards}")
        if search(hand, hand):
            rounds += 1
            continue
        if coin > 0 and search(hand, draw):
            rounds += 1
            coin -= 1
            continue
        if coin > 1 and search(draw, draw):
            rounds += 1
            coin -= 2
            continue
        
        break
    return rounds


def search(a, b):
    for k in a:
        if n + 1 - k in b:
            b.remove(n + 1 - k)
            a.remove(k)
            return True
    return False
            